### PR TITLE
fix(#7524): Open in New Tab action from a sub-object in a layout

### LINKED ISSUE #7524: Open in new tab has stopped working.
#### Summary
The "open in new tab" action from the context menu in display layouts has stopped working for some reason. It shows an error in the JavaScript console

#### Steps to Reproduce
1. Open (or create) a display layout containing a LAD table
2. Click on the context menu (three-dot menu) in the LAD table and select "Open In New Tab"
3. Observe that nothing happens and there is an error in the console

#### Environment
* Open MCT Version: 5f8d6899d229ffd8b466d8fca02c00bf67cbad3c
* Deployment Type: testathon

#### Impact Check List

- [ ] Data loss or misrepresented data?
- [x] Regression? Did this used to work or has it always been broken?
- [ ] Is there a workaround available?
- [x] Does this impact a critical component?
- [ ] Is this just a visual bug with no functional impact?
- [ ] Does this block the execution of e2e tests?
- [ ] Does this have an impact on Performance?

#### Additional Information

```
Uncaught TypeError: Cannot convert object to primitive value
    at paramsToArray (url.js:51:29)
    at objectPathToUrl (url.js:72:1)
    at OpenInNewTab.invoke (openInNewTabAction.js:35:30)
    at action.onItemClicked (MenuAPI.js:81:1)
    at callWithErrorHandling (runtime-core.esm-bundler.js:192:1)
    at callWithAsyncErrorHandling (runtime-core.esm-bundler.js:199:1)
    at HTMLLIElement.invoker (runtime-dom.esm-bundler.js:681:31)
    ```

### PR DESCRIPTION
**Pull request by ozyx**
Closes #7524
Closes https://github.com/nasa/openmct/issues/6982

### Describe your changes:

- Fixes the "Open in New Tab" action for sub-objects
    - Removes logic that was meant for "data pivoting" feature to be able to pass time bound info to the main Open MCT app. Since "data pivoting" was moved to core Open MCT, this is no longer needed and was just causing crashes.
- Greatly improves a11y of Menus and SuperMenus
- Add `.visibility-hidden` global utility CSS class for providing accessible descriptions for screen-readers without affecting the view
- Adds test for "Open in New Tab" from a display layout sub-object
- Adds test data generation for display layout w/ sub-plot
### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [ ] Is this a notable change that will require a special callout in the release notes [Notable Change](../docs/src/process/release.md) ? For example, will this break compatibility with existing APIs or projects which source these plugins?

### Author Checklist

* [x] Changes address original issue?
* [x] Tests included and/or updated with changes?
* [x] Has this been smoke tested?
* [x] Have you associated this PR with a `type:` label? Note: this is not necessarily the same as the original issue.
* [x] Have you associated a milestone with this PR? Note: leave blank if unsure.
* [ ] Is this a breaking change to be called out in the release notes?
* [ ] Testing instructions included in associated issue OR is this a dependency/testcase change?

### Reviewer Checklist

* [ ] Changes appear to address issue?
* [ ] Reviewer has tested changes by following the provided instructions?
* [ ] Changes appear not to be breaking changes?
* [ ] Appropriate automated tests included?
* [ ] Code style and in-line documentation are appropriate?

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -19,7 +19,7 @@
  this source code distribution or the Licensing information page available
  at runtime from the About dialog for additional information.
 -->
-<!DOCTYPE html>
+<!doctype html>
```
Comment (unlikelyzero) [inline index.html:22]: lint?

File: src/tools/urlSpec.js [hunk 2]
```diff
@@ -65,15 +65,5 @@ describe('the url tool', function () {
       const constructedURL = objectPathToUrl(openmct, mockObjectPath);
       expect(constructedURL).toContain('#/browse/mock-parent-folder/mock-folder');
     });
-    it('can take params to set a custom url', () => {
```
Comment (unlikelyzero) [inline src/tools/urlSpec.js:68]: do we have precise coverage for this

**Review by unlikelyzero — COMMENTED** [review]
Looks good so far. Let's take the time to understand what this means:

> Know When to Use aria-describedby
Sometimes, we need to give more information to an element. For example, we might want to tell the user that the button they will press will open a new tab.

This information is important because the user needs to know where they are when navigating websites.

For these types of scenarios, we can use aria-describedby to give additional information.

Example:

⛔ BAD

<button aria-label="Close" aria-label="Opens in a new tab">	
Show related		
</button>
✅ GOOD

<button aria-label="Close" aria-describedby="description">			Show related		
</button>		
<div id="description">Opens in a new tab</div>
In the first example above, what engineers expect the screen reader to announce is: “button, Show related, opens in a new tab”.

But the screen reader does not do that. Instead, it says,“button, opens in a new tab”. The screen reader does not read the content inside, because aria-label always overrides the text content of the HTML5 element it has been applied to.

The second code snipped shows the correct way to use aria-describedby. The screen reader will read, “button, Show related, opens in a new tab”.

That information tells the user that the button is to show related content and if they press that button, it will navigate them to another tab.

Takeaway: Use aria-describedby to add additional information to elements.

File: index.html (hunk 1 above)
Comment (ozyx) [reply index.html:22]: Yep

File: src/tools/urlSpec.js (hunk 2 above)
Comment (ozyx) [reply src/tools/urlSpec.js:68]: The code this tested was removed

File: e2e/appActions.js [hunk 3]
```diff
@@ -505,15 +505,14 @@ async function setTimeConductorBounds(page, startDate, endDate) {
...
+  // Activate Independent Time Conductor
+  await page.getByLabel('Enable Independent Time Conductor').click();
 
   // Bring up the time conductor popup
-  await page.click('.c-conductor-holder--compact .c-compact-tc');
+  await page.getByLabel('Independent Time Conductor Settings').click();
   await expect(page.locator('.itc-popout')).toBeInViewport();
-
-  await setTimeBounds(page, startDate, endDate);
+  await setTimeBounds(page, start, end);
```
Comment (unlikelyzero) [inline e2e/appActions.js:515]: setTimeBoundsOfITC ?

File: e2e/tests/framework/appActions.e2e.spec.js [hunk 4]
```diff
@@ -174,6 +174,6 @@ test.describe('AppActions', () => {
       type: 'Folder'
     });
     await openObjectTreeContextMenu(page, folder.url);
-    await expect(page.getByLabel('Menu')).toBeVisible();
+    await expect(page.getByLabel(`${folder.name} Context Menu`)).toBeVisible();
```
Comment (unlikelyzero) [inline e2e/tests/framework/appActions.e2e.spec.js:177]: woah

File: e2e/tests/functional/plugins/displayLayout/displayLayout.e2e.spec.js [hunk 5]
```diff
@@ -23,19 +23,128 @@ import { fileURLToPath } from 'url';
...
+    await page.getByLabel('Expand My Items folder').click();
+    const waitForMyItemsNavigation = page.waitForURL(`**/mine/?*`);
+    await page
+      .getByLabel('Main Tree')
+      .getByLabel('Navigate to Parent Display Layout layout Object')
+      .click();
+    // Wait for the URL to change to the display layout
+    await waitForMyItemsNavigation;
+  });
+  test('Open in New Tab action preserves time bounds', async ({ page }) => {
```
Comment (unlikelyzero) [inline e2e/tests/functional/plugins/displayLayout/displayLayout.e2e.spec.js:64]: anything that opens a second tab should use '@2p' as it's likely to use twice the memory when running tests in parallel on a CI agent and we may eventually have to run them in their own pipeline/job

File: e2e/tests/functional/plugins/displayLayout/displayLayout.e2e.spec.js [hunk 6]
```diff
@@ -23,19 +23,128 @@ import { fileURLToPath } from 'url';
...
+    // Open the Child Overlay Plot 1 in a new tab
+    await page.getByLabel('View menu items').click();
+    const pagePromise = page.context().waitForEvent('page');
+    await page.getByLabel('Open In New Tab').click();
+
+    const newPage = await pagePromise;
+    await newPage.waitForLoadState('domcontentloaded');
+
+    // Verify that the global time conductor bounds in the new page match the updated global bounds
+    expect(
```
Comment (unlikelyzero) [inline e2e/tests/functional/plugins/displayLayout/displayLayout.e2e.spec.js:123]: ok this makes sense

File: src/api/menu/components/SuperMenu.vue [hunk 7]
```diff
@@ -38,8 +38,8 @@
             :key="action.name"
             role="menuitem"
             :aria-disabled="action.isDisabled"
+            aria-describedby="item-description"
             :class="action.cssClass"
-            :title="action.description"
```
Comment (unlikelyzero) [inline src/api/menu/components/SuperMenu.vue:42]: I think we may want to keep titles, just not use them for tests

File: e2e/tests/functional/plugins/displayLayout/displayLayout.e2e.spec.js (hunk 5 above)
Comment (ozyx) [reply e2e/tests/functional/plugins/displayLayout/displayLayout.e2e.spec.js:64]: oh, I thought I added that. Maybe I added it to the suite by mistake

File: e2e/appActions.js (hunk 3 above)
Comment (ozyx) [reply e2e/appActions.js:515]: it seems like it was designed to work for any time conductor? idk, I think all of these methods need an overhaul. not in this PR though.

**Comment by ozyx** [comment]
Have a few more changes after talking to @davetsay . Moving back to draft for now

**Comment by davetsay** [comment]
Looking good so far, @ozyx . @unlikelyzero must be losing his mind over this one.

File: src/plugins/LADTable/components/LadRow.vue [hunk 8]
```diff
@@ -260,7 +260,7 @@ export default {
         event.preventDefault();
         this.preview(this.objectPath);
       } else {
-        const resultUrl = identifierToString(this.openmct, this.objectPath);
+        const resultUrl = objectPathToUrl(this.openmct, this.objectPath);
```
Comment (davetsay) [inline src/plugins/LADTable/components/LadRow.vue:263]: nice catch

Comment (ozyx) [reply src/plugins/LADTable/components/LadRow.vue:263]: right! we had a default export in addition to named exports in that file... scary

File: src/plugins/inspectorDataVisualization/TelemetryFrame.vue [hunk 9]
```diff
@@ -121,7 +121,8 @@ export default {
       };
       const newTabAction = this.openmct.actions.getAction('newTab');
       // No view context needed, so pass undefined.
-      // The urlParams arg will override the global time bounds with the dataviz plot bounds.
+      // The urlParams arg will override the global time bounds with the data visualization
```
Comment (davetsay) [inline src/plugins/inspectorDataVisualization/TelemetryFrame.vue:124]: haha

File: src/plugins/inspectorDataVisualization/TelemetryFrame.vue [hunk 10]
```diff
@@ -120,7 +120,9 @@ export default {
         'tc.mode': 'fixed'
       };
       const newTabAction = this.openmct.actions.getAction('newTab');
-      newTabAction.invoke([sourceTelemObject], urlParams);
+      // No view context needed, so pass undefined.
+      // The urlParams arg will override the global time bounds with the dataviz plot bounds.
+      newTabAction.invoke([sourceTelemObject], undefined, urlParams);
```
Comment (davetsay) [inline src/plugins/inspectorDataVisualization/TelemetryFrame.vue:126]: fantastic

File: src/plugins/openInNewTabAction/openInNewTabAction.js [hunk 11]
```diff
@@ -31,8 +31,26 @@ export default class OpenInNewTab {
 
     this._openmct = openmct;
   }
-  invoke(objectPath, urlParams = undefined) {
-    let url = objectPathToUrl(this._openmct, objectPath, urlParams);
+
```
Comment (davetsay) [inline src/plugins/openInNewTabAction/openInNewTabAction.js:34]: awesome

**Review by davetsay — CHANGES_REQUESTED** [review]
one question / potential change. and I'm going to smoke test this before merging.

File: src/plugins/inspectorDataVisualization/TelemetryFrame.vue [hunk 12]
```diff
@@ -70,17 +70,27 @@ export default {
...
+      domainObject: this.telemetryObject,
+      path: this.path,
+      renderWhenVisible: this.renderWhenVisible
     };
   },
   props: {
     bounds: {
       type: Object,
-      default: () => {}
+      required: true
```
[the remark this answers is not in the document]
Comment (ozyx) [reply src/plugins/inspectorDataVisualization/TelemetryFrame.vue]: reverted as we don't need this change

**Review by davetsay — APPROVED** [review]
Great work.

**Pull request merged**